### PR TITLE
Data cleaning and remove spelling mistake in outerlayer

### PR DESCRIPTION
**Pull request by Flogreeni**
also removed some testing parts and added new ones

### CODE REVIEW

**Review by maximilian-hildebrandt — CHANGES_REQUESTED** [review]
**General:**

- Overall, you are moving into the right direction of completing the recommendation outputs. Unfortunately, the code workbook suffers from some substantial issues that yet have not been resolved or completed. In general, the workbook lacks a detail check regarding overall structure, readability (many unnecessary functions, lack of comments), and completeness (e.g., evaluation missing; many feedback points from last review not integrated yet).
- I had good experiences when I scheduled time to review code diligently before submitting it, by going through the code conscientiously, assessing it from the reviewer perspective (Is it understandable? Is it complete? Is it structured?) and checking if I integrated all feedback sufficiently. 
- The chapter headings/structure require streamlining. Data collection code paragraphs are scattered throughout the notebook. Data visualization has no formal heading. Please structure into (1) Data collection & imports, (2) Data pre-processing & exploratory data analysis, (3) Model development, (4) Model evaluation, (5) Model deployment). 

**Data cleaning:**

- Assess frequency of missing values first. Is there any processing needed?
- (From last feedback) The rules seem arbitrary. Why did you choose those rules? All methodological choices require a rational argument. There are different strategies to handle missing values, you can for example impute via the mean or drop the observations with missing values. The more conservative approach is to drop observations with missing values. In our case I would suggest we output an error message in case of a missing value ("incomplete data"), since the app should provide data all the time. What do you think?

**Exploratory data analysis/data visualization:**
- Use a chapter heading "Exploratory data analysis".
- The exploratory data analysis is not sufficient (summary of data missing, data charts not applied correctly). Data cleaning is not really explained, arbitrary processing steps have been conducted (why replace missing gender with "d" for diverse?). 
- The chart selection to visualize data points is not yet reasonable:
- Scatter plot: Does not work with date information, since it requires interval-scale data. If you want to display data over time, pick a line or bar chart.
- Box plot: A box plot summarizes multiple data point by computing the median, quartiles, and outliers. Currently, your boxplot shows only one data point for each date, this does not work. Instead, you could show a box plot for all temperature data (i.e., one entry on the x axis for temperature). 
- Other visualizations that would be helpful: A bar chart showing frequency of the categories in heaviness of trip, bar chart of gender (x gender, y percentage of all observations), box plot for wind speed, bar chart of frequency of four weather states, a summary table (via df.describe())

**Model evaluation:**
- Generating the evaluation set is actually data collection, so move it into the first chapter on data collection.
- The actual evaluation is missing. You can read Matze's comments in as csv_file (in data collection), and compute the key evaluation metrics in this chapter. In the end, I would like to get a metric showing how precise the predictions were including an interpretation of the numbers.
- Details:
- Check spelling of header, punctuation missing, capital letter usage very random.
- Comment code for readability.
- Add random state to generate_evaluationset(), otherwise results look different each time you use the function.
- Instead of printing the generated set directly, turn it first into a dataframe for better readability.

**Deployment chapter:**
- Comments for code for readability. 
- You often output raw data to test your functions (e.g., "print(evaluation_user"). I suggest to always output dataframes with test functions, otherwise you get very long & visually unappealing text snippets that are hard to read.

**Comment by Flogreeni** [comment]
Hi Max,

pls review the new Notebook.
for better readability we kicked out a lot of this printing statements, which are just not helping at all to better understand the code. 
Tryed to at least have some explanation in front of every code section.

Last Open Point out of my view is a better visualisation of temperature through out a trip... but it is actually not really helpful, cause the way we are generating a dataset for recommendation purposes normal containts quit similar weather through out the day... so the charts just look stupid.

helped myself with this comment so far:
Data visualisation is not needed in this state of the model, it will be done later during the evaluation

cheers Flo

PS: sorry it is such a big request again for you to review... this is not really handy for reviewing... I understand the problem, also understand the solution with further branching, but kind of failed doing it...

**Review by maximilian-hildebrandt — APPROVED** [review]

**Comment by maximilian-hildebrandt** [comment]
General:

- The notebook improved substantially and is now in a good shape. It is well-structured, easy to read and the technical analyses are mostly correct. The only major mistake was related to the hyperparameter optimization on the test set (see details below). With the current status, it would already fulfill the requirements I would have. Any further tweaking is refinement for a very good notebook.
- Merge libraries and data collection chapters (No. 1 and No. 2). No need for an own chapter for libraries.
- Pull the dataframe you created in chapter No. 3 into the first section ("#Creating a dataframe"). This is the key output of the chapter that loads data and is not related to exploratory data analysis.

Exploratory data analysis:

- Round descriptive statistics to 2 digits for readability
- The data exploration now is quite nice! Good job. Some minor details: Only the gender plot seems to be incorrect, since there are not only male values in the plot. For box plots, swap axes, putting the temperature/wind speed on the y-axis.
- What you haven't covered so far is the distribution of the data. Adding Kernel density plots for windspeed and temperature would show you important other properties (shape of distribution, type of distribution).

Model development:

- Add to the initial comment an explanation how you came up with the formula "'temp + heatperception – (if female ==2) > current_temp'" or in other words, why this is a plausible formula for splitting the trees. Also add that you aligned the logic with the key product stakeholders to ensure a sound & realistic logic.

Model evaluation:

- You use the test set to optimize a hyperparameter (your heat preference). With that procedure, you actually "burn" your testset and need another dataset to achieve a plausible performance evaluation. You never want to use the test set for any optimization! Split your data into training, validation, and test set and never use the test set until you conduct the final evaluation result. (no need for any changes anymore, but rather a lesson for future projects).
- Remove all columns except for actual clothing layers from the evaluation plot. 
- Love the interpretation section! Great. But weren't there 1-2 small adjustments you had through conversation with Matze? If you changed anything, add it here.
